File: Python/257.Binary Tree Paths.py
#求出所有根节点到叶节点的路径
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def binaryTreePaths(self, root):
        if root==None:  return []
        self.result = []
        self.dfs(root,"")
        return self.result

    def dfs(self,node,path):
        if node.right==None and node.left==None:
            self.result.append(path+str(node.val))
            return
        if node.left!=None:
            self.dfs(node.left,path+str(node.val)+"->")
        if node.right!=None:
            self.dfs(node.right,path+str(node.val)+"->")


t1 = TreeNode(1)
t2 = TreeNode(2)
t3 = TreeNode(3)
t4 = TreeNode(4)
t5 = TreeNode(5)
t6 = TreeNode(6)
t7 = TreeNode(7)
t1.left = t2
t1.right = t3
t2.left = t4
t2.right = t6
t3.right = t5
t3.left = t7

s = Solution()
print(s.binaryTreePaths(t1))

File: C++/Tree/257.Binary Tree Paths.cpp
/*
求二叉树所有从根到叶结点的路径。
*/
/*
思路：递归将路径上的每个点加入即可，到根结点时将整个路径加入result数组。
*/
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
private:
	vector<string> result;
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        if(!root)	return result;
        helpBinaryTreePaths(root,to_string(root->val));
        return result;
    }
    void helpBinaryTreePaths(TreeNode* root,string prepath)
    {
    	if(!root->left && !root->right)	
    	{
    		result.push_back(prepath);
    		return ;
    	} 
    	if(root->left) helpBinaryTreePaths(root->left,prepath+"->"+to_string(root->left->val));
        if(root->right) helpBinaryTreePaths(root->right,prepath+"->"+to_string(root->right->val));
    }
};
int main()
{
	Solution s;
	TreeNode n1(1),n2(2),n3(3),n4(4),n5(5),n6(6);
	n4.left = &n3, n4.right = &n5;
	n3.left = &n1, n1.right = &n2;
	n5.right = &n6;
	vector<string> out = s.binaryTreePaths(&n4);
	for(string s:out)
		cout<<s<<endl;
}

